charmcraft/commands/build.py

```diff
@@ -371,11 +349,62 @@ def run(
                     bases_index,
                 )
 
-        if not charms:
+        return build_plan
+
+    def run(
+        self, bases_indices: Optional[List[int]] = None, destructive_mode: bool = False
+    ) -> List[str]:
+        """Run build process.
+
+        In managed-mode or destructive-mode, build for each bases configuration
+        which has a matching build-on to the host we are executing on.  Warn for
+        each base configuration that is incompatible.  Error if unable to
+        produce any builds for any bases configuration.
+
+        :returns: List of charm files created.
+        """
+        charms: List[str] = []
+
+        managed_mode = env.is_charmcraft_running_in_managed_mode()
+        if not managed_mode and not destructive_mode:
+            self.provider.ensure_provider_is_available()
+
+        if self.entrypoint:
+            notify_deprecation("dn04")
+
+        if self.requirement_paths:
+            notify_deprecation("dn05")
+
+        if not (self.charmdir / "charmcraft.yaml").exists():
+            notify_deprecation("dn02")
+
+        build_plan = self.plan(
+            bases_indices=bases_indices,
+            destructive_mode=destructive_mode,
+            managed_mode=managed_mode,
+        )
+        if not build_plan:
             raise CommandError(
                 "No suitable 'build-on' environment found in any 'bases' configuration."
             )
 
+        charms = []
+        for bases_config, build_on, bases_index, build_on_index in build_plan:
+            logger.debug(
+                "Building for 'bases[%d][%d]'.",
+                bases_index,
+                build_on_index,
```

Comment:
Maybe the arguments could be in a single line?

Reply:
done!